### PR TITLE
let builders create new imagestreams for pushes

### PR DESCRIPTION
**Pull request by deads2k**
Reminder to myself to add a test and make sure it works:

```
1. Login OpenShift and create a project;
 
2. Use the use token to docker login the docker-registry:
 
  docker login -u service-account-creds -p `oc whoami -t` -e any@example.com <integrated registry>
 
3. Pull Image and tag the image from ;
 
  docker pull busybox;
 
    docker tag busybox   <integrated registry>/projectname/imagestreamname:tag
 
4. Push the image to the OpenShift registry:
 
docker push <integrated registry>/projectname/imagestreamname:tag
 
 5. Pull the image in the OpenShift registry again:
 
  docker pull <integrated registry>/projectname/imagestreamname:tag
```

@pruan-rht I think its you.

### CODE REVIEW

File: pkg/cmd/server/bootstrappolicy/policy.go [hunk 1]
```diff
@@ -397,6 +397,8 @@ func GetBootstrapClusterRoles() []authorizationapi.ClusterRole {
 			Rules: []authorizationapi.PolicyRule{
 				// push and pull images
 				authorizationapi.NewRule("get", "update").Groups(imageGroup).Resources("imagestreams/layers").RuleOrDie(),
+				// allow auto-provisioning when pushing an image that doesn't have an imagestream yet
```
Comment (liggitt) [inline pkg/cmd/server/bootstrappolicy/policy.go]: do we always want builders to have this power?

Comment (deads2k) [inline pkg/cmd/server/bootstrappolicy/policy.go]: > do we always want builders to have this power?

I can't think of a reasonable reason not to.  If a cluster-admin comes up with one, he can change the default role.

**Comment by mfojtik** [comment]
@deads2k i would be nice to have something like `oc docker login` ;-) (just a wrapper that calls docker login, but fills all the required auth fields).

File: pkg/cmd/server/bootstrappolicy/policy.go (hunk 1 above)
Comment (mfojtik) [inline pkg/cmd/server/bootstrappolicy/policy.go]: does this allow custom builders to create imagestream for arbitrary images?

Comment (deads2k) [inline pkg/cmd/server/bootstrappolicy/policy.go]: > does this allow custom builders to create imagestream for arbitrary images?

Yes, but given push rights, they can destroy any existing one already.  The ability to create new ones doesn't do much if nothing is wired to use them.

**Comment by deads2k** [comment]
> @deads2k i would be nice to have something like oc docker login ;-) (just a wrapper that calls docker login, but fills all the required auth fields like auth, registry IP, etc.).

@juanvallejo Want to give that a try?  I'd want it to support `--as` as well.

@fabianofranz not sure what else you've got on his plate.

**Comment by kargakis** [comment]
yay!

**Comment by kargakis** [comment]
> @deads2k i would be nice to have something like oc docker login ;-) (just a wrapper that calls docker login, but fills all the required auth fields like auth, registry IP, etc.).

@aweiteka also wants this if I am not mistaken

**Comment by aweiteka** [comment]
> @aweiteka also wants this if I am not mistaken

+1. Yes, this would really streamline the dev experience. One blocking issue is we can't programmatically discover the registry endpoint. https://github.com/openshift/origin/issues/2162

Also if docker requires `sudo` that would be problematic knowing where to write the .dockercfg.

**Comment by deads2k** [comment]
test complete

**Comment by deads2k** [comment]
[test]

**Comment by deads2k** [comment]
[merge]

**Pull request merged**